testdata/forget_leader.txt

```diff
@@ -0,0 +1,423 @@
+log-level none
+----
+ok
+
+add-nodes 4 voters=(1,2,3) learners=(4) index=10
+----
+ok
+
+campaign 1
+----
+ok
+
+stabilize
+----
+ok (quiet)
+
+log-level debug
+----
+ok
+
+# ForgetLeader is a noop on the leader.
+forget-leader 1
+----
+ok
+
+raft-state
+----
+1: StateLeader (Voter) Term:1 Lead:1
+2: StateFollower (Voter) Term:1 Lead:1
+3: StateFollower (Voter) Term:1 Lead:1
+4: StateFollower (Non-Voter) Term:1 Lead:1
+
+# ForgetLeader causes a follower to forget its leader, but remain in the current
+# term. It's a noop if it's called again.
+forget-leader 2
+----
+INFO 2 forgetting leader 1 at term 1
+
+raft-state
+----
+1: StateLeader (Voter) Term:1 Lead:1
+2: StateFollower (Voter) Term:1 Lead:0
+3: StateFollower (Voter) Term:1 Lead:1
+4: StateFollower (Non-Voter) Term:1 Lead:1
+
+forget-leader 2
+----
+ok
+
+raft-state
+----
+1: StateLeader (Voter) Term:1 Lead:1
+2: StateFollower (Voter) Term:1 Lead:0
+3: StateFollower (Voter) Term:1 Lead:1
+4: StateFollower (Non-Voter) Term:1 Lead:1
+
+# ForgetLeader also works on learners.
+forget-leader 4
+----
+INFO 4 forgetting leader 1 at term 1
+
+raft-state
+----
+1: StateLeader (Voter) Term:1 Lead:1
+2: StateFollower (Voter) Term:1 Lead:0
+3: StateFollower (Voter) Term:1 Lead:1
+4: StateFollower (Non-Voter) Term:1 Lead:0
+
+# When receiving a heartbeat from the leader, they revert to followers.
+tick-heartbeat 1
```

Comment:
Since you've tested this I assume the answer is no, but in practice, do heartbeats from the (now removed) leader reverse-race with the call to `ForgetLeader`, leadering to a situation where the `ForgetLeader` is wiped out?

Reply:
Not in the cases we care about in CRDB, no.

For now, we'll be lenient and call this while processing (Pre)Vote requests, under lock, so we'll always grant the (Pre)Vote when appropriate. If we want to be more strict and only do this on unquiescence, then a delayed heartbeat would revert it to follower between a PreVote and Vote, causing it to reject the Vote. But that's ok, and sort of by design since we want to respect a current leader -- in the worst case, it will call another election after waiting out a timeout.